File: backend/nidp/services/daas_api/move_odds_profile.py
"""Move odds profile — each stock's quality beside its move odds (PRD "stock analysis rankings": movement probability
is not investment quality, so the two are shown side by side and never combined).

Pure functions only (grades, the ratio catalogue, event categories); the route that reads the database lives in
routers/move_odds.py. Every value served comes straight from an NIDP table, except two things computed here from real
rows and stated as such on the page:

  * the sector rating — the median V3 quality score of the sector's stocks whose score has at least 80% of its inputs;
  * the 1-year return — adjusted close 252 sessions apart. The stored stock_features_daily.return_252d_pct matches
    neither the raw nor the adjusted 252-session change (2026-09-16: TATACHEM stored +3.2%, computed −22.6%), so it is
    not served.

A missing value is null, never 0. A ratio is offered only when the universe actually has it (the metric-registry rule:
non-null for at least 5% of stocks and more than one distinct value); the rest are listed with the reason.
"""
from __future__ import annotations

from dataclasses import dataclass
from statistics import median
from typing import Any, Dict, Iterable, List, Optional, Sequence

GRADE_A = 70.0
GRADE_B = 50.0
COVERAGE_MIN = 80.0            # below this share of inputs a V3 score is marked partial and left out of sector medians
MIN_COVERAGE_PCT = 5.0         # a ratio is offered only if at least this share of the universe has it …
MIN_DISTINCT = 2               # … and it takes more than one value (177 readings of exactly 0.00 is not a ratio)

CAP_LABEL = {"LARGE_CAP": "Large", "MID_CAP": "Mid", "SMALL_CAP": "Small", "MICRO_CAP": "Micro"}


def grade(score: Optional[float]) -> Optional[str]:
    if score is None:
        return None
    return "A" if score >= GRADE_A else ("B" if score >= GRADE_B else "C")


def cap_label(bucket: Optional[str]) -> Optional[str]:
    return CAP_LABEL.get((bucket or "").upper()) if bucket else None


def _f(v: Any) -> Optional[float]:
    if v is None:
        return None
    try:
        x = float(v)
    except (TypeError, ValueError):
        return None
    return x if x == x else None          # NaN → None


def quality_block(row: Optional[Dict[str, Any]]) -> Optional[Dict[str, Any]]:
    """The V3 composite for one stock, or None when there is no score (never a default)."""
    if not row:
        return None
    score = _f(row.get("quality_score"))
    if score is None:
        return None
    cov = _f(row.get("quality_coverage_pct"))
    shown = round(score, 1)                                   # graded on the number shown beside it
    return {
        "score": shown, "grade": grade(shown), "coverage": round(cov, 1) if cov is not None else None,
        "partial": cov is None or cov < COVERAGE_MIN,
        "fundamental": _r1(_f(row.get("fundamental_score"))), "technical": _r1(_f(row.get("technical_score"))),
        "as_of": _iso(row.get("as_of_date")),
    }


def sector_ratings(scores: Iterable[Dict[str, Any]]) -> Dict[str, Dict[str, Any]]:
    """Median quality score per sector over stocks whose score has >= COVERAGE_MIN of its inputs, with the count.
    A sector with no such stock gets median None and grade None."""
    by: Dict[str, List[float]] = {}
    scored: Dict[str, int] = {}
    for r in scores:
        sec = r.get("sector")
        s = _f(r.get("quality_score"))
        if not sec or s is None:
            continue
        scored[sec] = scored.get(sec, 0) + 1
        cov = _f(r.get("quality_coverage_pct"))
        if cov is not None and cov >= COVERAGE_MIN:
            by.setdefault(sec, []).append(s)
    out = {}
    for sec, n_scored in scored.items():
        vals = by.get(sec, [])
        m = round(median(vals), 1) if vals else None
        out[sec] = {"median": m, "grade": grade(m), "n": len(vals), "n_scored": n_scored}
    return out


# ── Ratio catalogue ─────────────────────────────────────────────────────────────────────────────────────────────────
# Groups and the recent / preceding / historical columns follow the owner's design (a Screener-style picker). Only the
# `source` keys below are read; everything else in the design's list is shown as not available with its reason.
@dataclass(frozen=True)
class Ratio:
    key: str
    label: str
    group: str
    column: str                  # recent | preceding | historical
    unit: str                    # cr | pct | x | rs | pts
    source: Optional[str] = None  # column on the features row (or "return_1y", computed); None = not served
    note: Optional[str] = None
    reason: Optional[str] = None  # why it is not served, when source is None


_NOT_YET = "Held in NIDP but not served on this page yet"
_NOT_COMPUTED = "Not computed in NIDP"
_HISTORY = "Needs 3–5 years of filings; NIDP does not compute it"
_PRICE_DEPTH = "Price history on record is under 3 years"
_ZERO_NOTE = "A reading of 0 can mean none or no data; the two are not told apart"

CATALOGUE: List[Ratio] = [
    Ratio("sales_ttm", "Sales (TTM)", "Most Used", "recent", "cr", reason=_NOT_YET),
    Ratio("opm", "Operating margin (OPM)", "Most Used", "recent", "pct", "operating_margin_pct"),
    Ratio("pat_ttm", "Profit after tax (TTM)", "Most Used", "recent", "cr", reason=_NOT_YET),
    Ratio("mcap", "Market capitalisation", "Most Used", "recent", "cr", "market_cap_cr"),
    Ratio("sales_q", "Sales latest quarter", "Quarter P&L", "recent", "cr", reason=_NOT_YET),
    Ratio("pat_q", "Profit after tax latest quarter", "Quarter P&L", "recent", "cr", reason=_NOT_YET),
    Ratio("sales_q_yoy", "YoY quarterly sales growth", "Quarter P&L", "recent", "pct", reason=_NOT_YET),
    Ratio("pat_q_yoy", "YoY quarterly profit growth", "Quarter P&L", "recent", "pct", reason=_NOT_YET),
    Ratio("pe", "Price to earnings", "Ratios", "recent", "x", "pe_ttm"),
    Ratio("divy", "Dividend yield", "Ratios", "recent", "pct", "dividend_yield_pct", note=_ZERO_NOTE),
    Ratio("pbv", "Price to book value", "Ratios", "recent", "x", "pb"),
    Ratio("roce", "Return on capital employed", "Ratios", "recent", "pct", "roce_pct"),
    Ratio("roe", "Return on equity", "Ratios", "recent", "pct", "roe_pct"),
    Ratio("npm", "Net profit margin", "Ratios", "recent", "pct", "profit_margin_pct"),
    Ratio("evebitda", "EV / EBITDA", "Ratios", "recent", "x", "ev_ebitda"),
    Ratio("ev", "Enterprise value", "Ratios", "recent", "cr", "enterprise_value_cr"),
    Ratio("roa", "Return on assets", "Ratios", "recent", "pct", reason=_NOT_COMPUTED),
    Ratio("ey", "Earnings yield", "Ratios", "recent", "pct", reason=_NOT_COMPUTED),
    Ratio("indpe", "Industry P/E", "Ratios", "recent", "x", reason="Not computed in NIDP (the sector median P/E column is empty)"),
    Ratio("ps", "Price to sales", "Ratios", "recent", "x", reason=_NOT_COMPUTED),
    Ratio("icr", "Interest coverage", "Ratios", "preceding", "x", "interest_coverage"),
    Ratio("peg", "PEG ratio", "Ratios", "preceding", "x", reason=_NOT_COMPUTED),
    Ratio("de", "Debt to equity", "Balance Sheet", "recent", "x", "debt_to_equity"),
    Ratio("debt", "Debt", "Balance Sheet", "recent", "cr", reason=_NOT_YET),
    Ratio("prom", "Promoter holding", "Balance Sheet", "recent", "pct", "promoter_pct"),
    Ratio("dprom", "Change in promoter holding (QoQ)", "Balance Sheet", "recent", "pts", "promoter_pct_change_qoq"),
    Ratio("pledge", "Pledged percentage", "Balance Sheet", "recent", "pct", "promoter_pledged_pct", note=_ZERO_NOTE),
    Ratio("cr", "Current ratio", "Balance Sheet", "preceding", "x", "current_ratio"),
    Ratio("eps", "EPS (TTM)", "Annual P&L", "recent", "rs", reason=_NOT_YET),
    Ratio("sg", "Sales growth (TTM, YoY)", "Annual P&L", "preceding", "pct", "revenue_growth_yoy_pct"),
    Ratio("pg", "Profit growth (TTM, YoY)", "Annual P&L", "preceding", "pct", "pat_growth_yoy_pct"),
    Ratio("epsg", "EPS growth (TTM, YoY)", "Annual P&L", "preceding", "pct", "eps_growth_yoy_pct"),
    Ratio("sg3", "Sales growth 3 years", "Annual P&L", "historical", "pct", reason=_HISTORY),
    Ratio("sg5", "Sales growth 5 years", "Annual P&L", "historical", "pct", reason=_HISTORY),
    Ratio("pg3", "Profit growth 3 years", "Annual P&L", "historical", "pct", reason=_HISTORY),
    Ratio("pg5", "Profit growth 5 years", "Annual P&L", "historical", "pct", reason=_HISTORY),
    Ratio("aroe3", "Average return on equity 3 years", "Annual P&L", "historical", "pct", reason=_HISTORY),
    Ratio("cfopat", "Cash from operations / profit", "Cash Flow", "recent", "x", "cfo_pat_ratio"),
    Ratio("pfcf", "Price to free cash flow", "Cash Flow", "recent", "x", reason=_NOT_COMPUTED),
    Ratio("close", "Close", "Price", "recent", "rs", "close"),
    Ratio("r1m", "Return over 1 month (20 sessions)", "Price", "historical", "pct", "return_20d_pct"),
    Ratio("r3m", "Return over 3 months (60 sessions)", "Price", "historical", "pct", "return_60d_pct"),
    Ratio("r1y", "Return over 1 year (252 sessions, adjusted)", "Price", "historical", "pct", "return_1y",
          note="Adjusted closes 252 sessions apart; adjusted for the corporate actions on record"),
    Ratio("r3y", "Return over 3 years", "Price", "historical", "pct", reason=_PRICE_DEPTH),
]
FEATURE_COLUMNS: List[str] = sorted({r.source for r in CATALOGUE if r.source and r.source != "return_1y"})
GROUPS = ["Most Used", "Annual P&L", "Quarter P&L", "Balance Sheet", "Cash Flow", "Ratios", "Price", "User Ratios"]


def measure(values: Sequence[Optional[float]], universe: int) -> Dict[str, Any]:
    nn = [v for v in values if v is not None]
    pct = (100.0 * len(nn) / universe) if universe else 0.0
    return {"n": len(nn), "covered_pct": round(pct, 1), "distinct": len(set(nn))}


def offered(m: Dict[str, Any]) -> bool:
    return m["covered_pct"] >= MIN_COVERAGE_PCT and m["distinct"] >= MIN_DISTINCT


def build_catalogue(values_by_key: Dict[str, List[Optional[float]]], universe: int) -> tuple[List[Dict[str, Any]], List[str]]:
    """(catalogue payload, keys of the offered ratios in catalogue order). values_by_key holds one list per served key,
    aligned to the universe; unserved keys are absent."""
    cat, keys = [], []
    for r in CATALOGUE:
        entry = {"key": r.key, "label": r.label, "group": r.group, "column": r.column, "unit": r.unit, "note": r.note}
        if r.source is None:
            entry.update(available=False, n=0, reason=r.reason)
        else:
            m = measure(values_by_key.get(r.key, []), universe)
            ok = offered(m)
            entry.update(available=ok, n=m["n"], covered_pct=m["covered_pct"],
                         reason=None if ok else ("Every reading is identical" if m["distinct"] == 1 else
                                                 f"Only {m['covered_pct']:.1f}% of these stocks have it (needs {MIN_COVERAGE_PCT:.0f}%)"))
            if ok:
                keys.append(r.key)
        cat.append(entry)
    return cat, keys


# ── Events → categories ─────────────────────────────────────────────────────────────────────────────────────────────
# The published events carry the catalyst engine's (type, subtype). They are mapped onto the announcement classifier's
# category names (metric_registry.EVENT_CATEGORIES) so the page offers one vocabulary. "qip" is shown as "Fund raise":
# fund_raise also covers rights and preferential issues, so calling all of it QIP would be wrong.
EVENT_LABELS: Dict[str, str] = {
    "management": "Management", "dividend": "Dividend", "qip": "Fund raise", "mna": "M&A", "rating": "Rating",
    "orders": "Orders", "earnings": "Earnings", "litigation": "Litigation", "capex": "Capex", "buyback": "Buyback",
    "regulatory": "Regulatory", "other": "Other",
}
_MANAGEMENT = {"director_change", "resignation", "management_change", "appointment"}
_CORP_MNA = {"acquisition", "divestment", "restructuring", "merger", "demerger", "stake_sale"}


def event_category(event_type: Optional[str], subtype: Optional[str]) -> str:
    t, s = (event_type or "").upper(), (subtype or "").lower()
    if t == "RESULTS":
        return "earnings"
    if t == "CAPITAL":
        return {"dividend": "dividend", "buyback": "buyback", "fund_raise": "qip"}.get(s, "other")
    if t == "M&A":
        return "mna"
    if t == "CORPORATE":
        if s in _MANAGEMENT:
            return "management"
        if s in _CORP_MNA:
            return "mna"
        if s in {"capacity", "expansion", "capex"}:
            return "capex"
        return "other"
    if t == "FINANCIAL":
        return "rating" if s.startswith("rating") else "other"
    if t == "CONTRACT":
        return "orders"
    if t == "GOVERNMENT":
        return "orders" if s in {"tender", "procurement", "project"} else "regulatory"
    if t == "INSTITUTION":
        return "management" if s == "appointment" else "other"
    if t == "LEGAL":
        return "litigation"
    if t in {"REGULATORY", "PHARMA"}:
        return "regulatory"
    return "other"


def events_block(events: Iterable[Dict[str, Any]]) -> Dict[str, Any]:
    """Count, material count (classified positive, negative or mixed), the latest time, and categories by count."""
    n = material = 0
    latest = None
    counts: Dict[str, int] = {}
    for e in events:
        n += 1
        if (e.get("direction") or "").lower() in {"positive", "negative", "mixed"}:
            material += 1
        t = e.get("event_time")
        if t is not None and (latest is None or t > latest):
            latest = t
        c = event_category(e.get("event_type"), e.get("event_subtype"))
        counts[c] = counts.get(c, 0) + 1
    cats = sorted(counts, key=lambda c: (-counts[c], list(EVENT_LABELS).index(c)))
    return {"n": n, "material": material, "latest": _iso(latest), "categories": cats}


def _r1(v: Optional[float]) -> Optional[float]:
    return None if v is None else round(v, 1)


def _iso(v):
    return v.isoformat() if hasattr(v, "isoformat") else v

File: frontend-v5/src/pages/Research/moveOddsProfile.ts
/**
 * Move odds · profile helpers (owner's design of 2026-09-18, continuation of the stock-analysis-rankings PRD).
 *
 * Pure functions over the /api/move-odds/profile payload: ratio formatting, filter conditions, grade tones and the six
 * question answers in the stock view. Rules they keep:
 *   · Quality is shown BESIDE movement and never combined with it (PRD: "distinguish movement probability from
 *     investment quality").
 *   · A missing value reads "not on record" — never 0 — and a row without a value for an active condition is
 *     excluded, not waved through.
 *   · No recommendation words (owner decision: scores only; D2 unchanged).
 */
import type { MoveHead, MoveProfile, MoveProfileRow, MoveRatioDef } from "@/services/adapters/moveOdds.adapter";

export type Cap = "All" | "Large" | "Mid" | "Small" | "Micro";
export const CAPS: Cap[] = ["All", "Large", "Mid", "Small", "Micro"];
export type Tone = "mint" | "amber" | "danger" | "ink";
export type Cond = { op: ">" | "<"; val: string };          // val as typed; the condition applies only once it is a number

// Short names for the ratios shown under a stock's symbol.
export const SHORT: Record<string, string> = {
  mcap: "MCap", pe: "P/E", pbv: "P/BV", roce: "ROCE", roe: "ROE", opm: "OPM", npm: "NPM", de: "D/E", divy: "Div yield",
  prom: "Promoter", dprom: "Δ promoter", pledge: "Pledged", cr: "Current", sg: "Sales gr.", pg: "Profit gr.", epsg: "EPS gr.",
  evebitda: "EV/EBITDA", ev: "EV", icr: "Int. cover", cfopat: "CFO/PAT", close: "Close", r1m: "1M", r3m: "3M", r1y: "1Y",
};
// Suggested thresholds shown as placeholders only; nothing filters until a value is typed.
export const SUGGEST: Record<string, [">" | "<", number]> = {
  mcap: [">", 1000], opm: [">", 10], roce: [">", 15], roe: [">", 12], de: ["<", 0.5], pe: ["<", 30], pbv: ["<", 5], divy: [">", 0.5],
  prom: [">", 50], pledge: ["<", 5], sg: [">", 10], pg: [">", 10], r1y: [">", 0], r3m: [">", 0], icr: [">", 3], cr: [">", 1],
};

const INR = new Intl.NumberFormat("en-IN", { maximumFractionDigits: 0 });
const INR2 = new Intl.NumberFormat("en-IN", { minimumFractionDigits: 2, maximumFractionDigits: 2 });

export function fmtRatio(unit: MoveRatioDef["unit"], v: number | null | undefined): string {
  if (v == null || !isFinite(v)) return "—";
  switch (unit) {
    case "cr": return `₹${INR.format(v)} Cr`;
    case "pct": return `${v.toFixed(1)}%`;
    case "x": return `${v.toFixed(1)}×`;
    case "rs": return `₹${INR2.format(v)}`;
    case "pts": return `${v > 0 ? "+" : v < 0 ? "−" : ""}${Math.abs(v).toFixed(2)} pts`;
  }
}

export function ratioIndex(p: MoveProfile): Record<string, number> {
  return Object.fromEntries(p.ratio_keys.map((k, i) => [k, i]));
}

export function valueOf(p: MoveProfile, idx: Record<string, number>, sym: string, key: string): number | null {
  const i = idx[key];
  if (i === undefined) return null;
  const v = p.rows[sym]?.ratios[i];
  return v == null ? null : v;
}

export function condActive(c: Cond | undefined): boolean {
  if (!c) return false;
  const t = c.val.trim();
  return t !== "" && isFinite(Number(t));
}

/** "pass" | "fail" | "missing": a stock without a value for an active condition is "missing" and is excluded. */
export function checkConds(p: MoveProfile, idx: Record<string, number>, sym: string, conds: Record<string, Cond>): "pass" | "fail" | "missing" {
  let missing = false;
  for (const [key, c] of Object.entries(conds)) {
    if (!condActive(c)) continue;
    const v = valueOf(p, idx, sym, key);
    if (v == null) { missing = true; continue; }
    const n = Number(c.val);
    if (c.op === ">" ? !(v > n) : !(v < n)) return "fail";
  }
  return missing ? "missing" : "pass";
}

export function gradeTone(g: string | null | undefined): Tone {
  return g === "A" ? "mint" : g === "B" ? "amber" : g === "C" ? "danger" : "ink";
}

// The copilot's stock card bands (backend/services/copilot_tools/instrument_research.py) so both cards read alike.
export function fundBand(s: number): { label: string; tone: Tone } {
  return s >= 75 ? { label: "Strong", tone: "mint" } : s >= 58 ? { label: "Above average", tone: "mint" }
    : s >= 45 ? { label: "Average", tone: "amber" } : { label: "Below average", tone: "danger" };
}
export function techBand(s: number): { label: string; tone: Tone } {
  return s >= 60 ? { label: "Bullish", tone: "mint" } : s >= 45 ? { label: "Neutral", tone: "amber" } : { label: "Bearish", tone: "danger" };
}
export function qualityLabel(s: number): string {
  return s >= 75 ? "Strong" : s >= 60 ? "Good" : s >= 45 ? "Average" : "Weak";
}

function median(xs: number[]): number | null {
  if (!xs.length) return null;
  const s = [...xs].sort((a, b) => a - b), m = Math.floor(s.length / 2);
  return s.length % 2 ? s[m] : (s[m - 1] + s[m]) / 2;
}

/** Median of a ratio across this run's scored stocks (about 1,000, not just the visible page) that share a sector and
 *  have the value, the stock itself included — stated with n. */
export function peerMedian(p: MoveProfile, idx: Record<string, number>, sector: string | null, key: string): { median: number | null; n: number } {
  if (!sector) return { median: null, n: 0 };
  const vals: number[] = [];
  for (const [sym, r] of Object.entries(p.rows)) {
    if (r.sector !== sector) continue;
    const v = valueOf(p, idx, sym, key);
    if (v != null) vals.push(v);
  }
  return { median: median(vals), n: vals.length };
}

export type Bullet = { tone: Tone; text: string };
export type Answer = { label: string; lead: string; bullets: Bullet[] };

const signedPct = (v: number) => `${v > 0 ? "+" : v < 0 ? "−" : ""}${Math.abs(v).toFixed(1)}%`;
const tone = (v: number | null) => (v == null ? "ink" : v > 0 ? "mint" : v < 0 ? "danger" : "ink") as Tone;
const P = (x: number | null | undefined) => (x == null ? "—" : `${(x * 100).toFixed(1)}%`);

export type AnswerCtx = {
  sym: string;
  sector: string | null;
  profile: MoveProfile;
  row: MoveProfileRow | null;
  idx: Record<string, number>;
  estimates: Partial<Record<MoveHead, number>>;
  baseRates: Partial<Record<MoveHead, number>>;
  day: (iso: string | null | undefined) => string;
  eventLabel: (key: string) => string;
};

/** The six question chips. Every sentence is built from served numbers; a missing number says so. */
export function answers(c: AnswerCtx): Answer[] {
  const { profile: p, row, idx, sym } = c;
  const v = (k: string) => valueOf(p, idx, sym, k);
  const unit = (k: string) => p.catalogue.find((d) => d.key === k)?.unit ?? "x";
  const f = (k: string) => fmtRatio(unit(k), v(k));
  const q = row?.quality ?? null;
  const sec = row?.sector ? p.sectors[row.sector] ?? null : null;
  const pe = peerMedian(p, idx, row?.sector ?? c.sector, "pe");
  const roce = peerMedian(p, idx, row?.sector ?? c.sector, "roce");
  const secName = row?.sector ?? c.sector ?? "its sector";
  const out: Answer[] = [];

  // 1 · Quality in brief
  if (!q) {
    out.push({ label: "Quality in brief", lead: `No V3 quality score is on record for ${sym} as of ${c.day(p.scores_as_of)}, so it has no rating.`,
      bullets: [{ tone: "ink", text: "Ratings are shown only where a score exists; a missing one is never filled in." }] });
  } else {
    const fb = q.fundamental != null ? `fundamentals ${fundBand(q.fundamental).label.toLowerCase()} (${Math.round(q.fundamental)}/100)` : "fundamentals not scored";
    const tb = q.technical != null ? `technicals ${techBand(q.technical).label.toLowerCase()} (${Math.round(q.technical)}/100)` : "technicals not scored";
    const sg = v("sg"), opm = v("opm"), de = v("de"), prom = v("prom");
    out.push({
      label: "Quality in brief",
      lead: `Rated ${q.grade} (${q.score.toFixed(1)}/100) as of ${c.day(q.as_of)}${q.partial ? `, from only ${q.coverage ?? "—"}% of its inputs` : ""}: ${fb}, ${tb}.`,
      bullets: [
        { tone: tone(sg), text: sg == null ? `Sales growth over twelve months: not on record. OPM ${f("opm")}.` : `Sales ${sg >= 0 ? "grew" : "fell"} ${Math.abs(sg).toFixed(1)}% over the last twelve months against the twelve before; OPM ${opm == null ? "not on record" : f("opm")}.` },
        { tone: de == null ? "ink" : de < 1 ? "mint" : "amber", text: `Debt to equity ${de == null ? "not on record" : f("de")}; promoter holding ${prom == null ? "not on record" : f("prom")}.` },
      ],
    });
  }

  // 2 · How's it performed?
  const r1y = v("r1y"), r3m = v("r3m"), r1m = v("r1m"), roe = v("roe"), pg = v("pg");
  const win = p.return_1y_window;
  out.push({
    label: "How's it performed?",
    lead: (r1y == null ? "No 1-year return on record (it needs 253 sessions of adjusted prices)." :
      `Over the year to ${c.day(win?.to)} it is ${r1y >= 0 ? "up" : "down"} ${Math.abs(r1y).toFixed(1)}% on adjusted closes (from ${c.day(win?.from)}).`) +
      ` Three months ${r3m == null ? "—" : signedPct(r3m)}, one month ${r1m == null ? "—" : signedPct(r1m)}.`,
    bullets: [
      { tone: tone(r1y), text: `Return on equity ${roe == null ? "not on record" : f("roe")}; return on capital employed ${v("roce") == null ? "not on record" : f("roce")}.` },
      { tone: tone(pg), text: pg == null ? "Profit growth over twelve months: not on record." : `Profit ${pg >= 0 ? "grew" : "fell"} ${Math.abs(pg).toFixed(1)}% over the last twelve months against the twelve before.` },
    ],
  });

  // 3 · Cheap or pricey?
  const peV = v("pe");
  out.push({
    label: "Cheap or pricey?",
    lead: (peV == null ? "No P/E on record (a loss, or no profit figure)." :
      `P/E ${f("pe")}${pe.median != null && pe.n > 1 ? ` against a median of ${pe.median.toFixed(1)}× for the ${pe.n} ${secName} stocks scored in this run with a P/E` : ""}.`) +
      ` Price to book ${f("pbv")}.`,
    bullets: [
      { tone: "ink", text: `EV / EBITDA ${f("evebitda")}; market capitalisation ${f("mcap")}.` },
      { tone: "ink", text: `Dividend yield ${f("divy")}${v("divy") === 0 ? " (a 0 can mean no dividend or no data)" : ""}.` },
    ],
  });

  // 4 · How risky is it?
  const e = c.estimates, b = c.baseRates;
  const pledge = v("pledge");
  out.push({
    label: "How risky is it?",
    lead: `Next session: ${P(e.p_up10_1d)} chance of touching +10% and ${P(e.p_down10_1d)} of −10% (base rates ${P(b.p_up10_1d)} and ${P(b.p_down10_1d)}); for 5%, ${P(e.p_up5_1d)} up and ${P(e.p_down5_1d)} down. These are estimates of a touch, not of the close.`,
    bullets: [
      { tone: v("de") == null ? "ink" : (v("de") as number) < 1 ? "mint" : "amber", text: `Debt to equity ${f("de")}; interest coverage ${f("icr")}.` },
      { tone: pledge != null && pledge > 0 ? "danger" : "ink",
        text: `${row?.cap ? `${row.cap} cap` : "Market-cap bucket not on record"}; promoter shares pledged ${pledge == null ? "not on record" : f("pledge")}${pledge === 0 ? " (a 0 can mean none or no data)" : ""}.` },
    ],
  });

  // 5 · Compare to peers
  const diff = q && sec?.median != null ? q.score - sec.median : null;
  const rc = v("roce");
  out.push({
    label: "Compare to peers",
    lead: sec?.median == null ? `No sector rating for ${secName}: none of its scored stocks has at least ${p.coverage_min}% of its inputs.` :
      `${secName} rates ${sec.grade} (${sec.median.toFixed(1)}/100, the median of ${sec.n} stocks with at least ${p.coverage_min}% of their inputs)` +
      (diff == null ? "; this stock has no score." : `; this stock is ${Math.abs(diff).toFixed(1)} points ${diff >= 0 ? "above" : "below"} it.`),
    bullets: [
      { tone: diff == null ? "ink" : diff >= 0 ? "mint" : "danger", text: pe.median != null && peV != null ? `P/E ${f("pe")} against ${pe.median.toFixed(1)}× for ${pe.n} ${secName} stocks scored in this run.` : "P/E comparison: not on record for this stock or its peers." },
      { tone: "ink", text: roce.median != null && rc != null ? `ROCE ${f("roce")} against a median of ${roce.median.toFixed(1)}% for ${roce.n} ${secName} stocks scored in this run.` : "ROCE comparison: not on record for this stock or its peers." },
    ],
  });

  // 6 · Any events?
  const ev = row?.events;
  out.push({
    label: "Any events?",
    lead: !ev || ev.n === 0 ? "No classified filings or reports on record in the five days to the freeze." :
      `${ev.n} on record — ${ev.categories.map(c.eventLabel).join(", ")}; latest ${ev.latest ? c.day(ev.latest) : "—"}.`,
    bullets: [
      { tone: ev && ev.material > 0 ? "amber" : "ink", text: `${ev?.material ?? 0} classified positive, negative or mixed (material); the rest neutral.` },
      { tone: "ink", text: "Events are listed for context; they are not model inputs. The full list is below." },
    ],
  });
  return out;
}
